include/envoy/thread_local/thread_local.h

```diff
@@ -16,49 +16,29 @@ namespace ThreadLocal {
 class ThreadLocalObject {
 public:
   virtual ~ThreadLocalObject() {}
-
-  /**
-   * The owning thread is about to exit. Cleanup any variables that may reference other thread
-   * locals.
-   */
-  virtual void shutdown() PURE;
 };
 
 typedef std::shared_ptr<ThreadLocalObject> ThreadLocalObjectSharedPtr;
```

Comment:
While reading this PR, I kind of wished that C++ had a variant of `shared_ptr` that would enforce (maybe just via ASSERTs like we do in `Dispatcher`) thread locality/affinity. I.e. it would catch leaking of the pointer across threads.